lib/kasan/kasan.c

```diff
@@ -0,0 +1,340 @@
+#endif
+
+#include "kasan_internal.h"
+#include <uk/kasan.h>
+
+static int kasan_ready;
+
+static const char *code_name(uint8_t code)
+{
+	switch (code) {
+	case KASAN_CODE_STACK_LEFT:
+	case KASAN_CODE_STACK_MID:
+	case KASAN_CODE_STACK_RIGHT:
+		return "stack buffer-overflow";
+	case KASAN_CODE_GLOBAL_OVERFLOW:
+		return "global buffer-overflow";
+	case KASAN_CODE_KMEM_FREED:
+		return "kmem use-after-free";
+	case KASAN_CODE_POOL_OVERFLOW:
+		return "pool buffer-overflow";
+	case KASAN_CODE_POOL_FREED:
+		return "pool use-after-free";
+	case KASAN_CODE_KMALLOC_OVERFLOW:
+		return "buffer-overflow";
+	case KASAN_CODE_KMALLOC_FREED:
+		return "use-after-free";
+	case 1 ... 7:
+		return "partial redzone";
+	default:
+		return "unknown redzone";
+	}
+}
+
+/* Check whether all bytes from range [addr, addr + size) are mapped to
+ * a single shadow byte
+ */
+static inline bool
+access_within_shadow_byte(uintptr_t addr, size_t size) {
+	return (addr >> KASAN_SHADOW_SCALE_SHIFT) ==
+		((addr + size - 1) >> KASAN_SHADOW_SCALE_SHIFT);
+}
+
+static inline bool
+shadow_1byte_isvalid(uintptr_t addr, uint8_t *code) {
+	int8_t shadow_val = (int8_t)*kasan_md_addr_to_shad(addr);
+	int8_t last = addr & KASAN_SHADOW_MASK;
+
+	if (likely(shadow_val == 0 || last < shadow_val))
+		return true;
+	*code = shadow_val;
+	return false;
+}
+
+static inline bool
+shadow_2byte_isvalid(uintptr_t addr, uint8_t *code) {
+	if (!access_within_shadow_byte(addr, 2))
+		return shadow_1byte_isvalid(addr, code) &&
+			shadow_1byte_isvalid(addr + 1, code);
+
+	int8_t shadow_val = *kasan_md_addr_to_shad(addr);
+	int8_t last = (addr + 1) & KASAN_SHADOW_MASK;
+
+	if (likely(shadow_val == 0 || last < shadow_val))
+		return true;
+	*code = shadow_val;
+	return false;
+}
+
+static inline bool
+shadow_4byte_isvalid(uintptr_t addr, uint8_t *code) {
+	if (!access_within_shadow_byte(addr, 4))
+		return shadow_2byte_isvalid(addr, code) &&
+			shadow_2byte_isvalid(addr + 2, code);
+
+	int8_t shadow_val = *kasan_md_addr_to_shad(addr);
+	int8_t last = (addr + 3) & KASAN_SHADOW_MASK;
+
+	if (likely(shadow_val == 0 || last < shadow_val))
+		return true;
+	*code = shadow_val;
+	return false;
+}
+
+static inline bool
+shadow_8byte_isvalid(uintptr_t addr, uint8_t *code)
+{
+	if (!access_within_shadow_byte(addr, 8))
+		return shadow_4byte_isvalid(addr, code) &&
+			shadow_4byte_isvalid(addr + 4, code);
+
+	int8_t shadow_val = *kasan_md_addr_to_shad(addr);
+	int8_t last = (addr + 7) & KASAN_SHADOW_MASK;
+
+	if (likely(shadow_val == 0 || last < shadow_val))
+		return true;
+	*code = shadow_val;
+	return false;
+}
+
+static inline bool
+shadow_Nbyte_isvalid(uintptr_t addr, size_t size, uint8_t *code)
+{
+		for (size_t i = 0; i < size; i++)
+			if (unlikely(!shadow_1byte_isvalid(addr + i, code)))
+				return false;
+		return true;
+}
+
+static inline void
+shadow_check(uintptr_t addr, size_t size, bool read)
+{
+	if (unlikely(!kasan_ready))
+		return;
+	if (unlikely(!kasan_md_addr_supported(addr)))
```

Comment:
Is also a check for `kasan_md_addr_supported(addr + size)` necessary?

Reply:
I don't think this is necessary, what make you think so?